ledger/src/blockstore/blockstore_purge.rs

```diff
@@ -239,6 +261,31 @@ impl Blockstore {
         }
         delete_range_timer.stop();
 
+        if purge_type == PurgeType::ExactAndCleanupChaining {
+            let mut slot_meta = slot_meta.expect("Slot meta should be present by this point");
+            let slot = from_slot; // also equal to to_slot
+            if let Some(parent_slot) = slot_meta.parent_slot {
+                let parent_slot_meta = self.meta(parent_slot)?;
+                if let Some(mut parent_slot_meta) = parent_slot_meta {
+                    // .retain() is a linear scan; however, next_slots should
+                    // only contain several elements so this isn't so bad
+                    parent_slot_meta
+                        .next_slots
+                        .retain(|&next_slot| next_slot != slot);
+                    write_batch.put::<cf::SlotMeta>(parent_slot, &parent_slot_meta)?;
+                } else {
+                    error!("Parent slot meta {} for child {} is missing. In the absence of a duplicate block this
+                        likely means a cluster restart was performed and your node contains invalid shreds generated
+                        with the wrong shred version, whose ancestors have been cleaned up.
+                        Falling back to duplicate block handling to remedy the situation", parent_slot, slot);
+                }
+            }
+
+            // Reinsert parts of `slot_meta` that are important to retain, like the `next_slots` field.
+            slot_meta.clear_unconfirmed_slot();
+            write_batch.put::<cf::SlotMeta>(slot, &slot_meta)?;
+        }
+
```

Comment:
This and the above together seem like enough unique logic to warrant a separate function. I think it would make sense to:

1.  move this into a separate function that takes a write_batch and a range https://github.com/solana-labs/solana/blob/45e292d8214f84972c7cc798a43190ec6cbf1c0c/ledger/src/blockstore/blockstore_purge.rs#L186-L253
2. Make a `purge_single_slot(slot: slot)` function that makes a write batch and moves it into 1.

Reply:
I was planning to bring this up for a future pr: I think actually this logic should be always be present? 
Right now what we're doing is ensuring that if we purge a single slot from a fork, the repair of that slot will allow replay to smoothly process the fork. 

However for a regular purge:
`4 - 5 - 6 - 7 - 8 - 9 - 10`
if we purge `Exact` for `from_slot = 5, to_slot = 8` then I don't believe we'll ever be able to replay slots `9, 10` on a successful repair. 
it seems we should extend the logic so that:
```
 parent_slot_meta
      .next_slots
      .retain(|&next_slot| next_slot != slot);
```
is **always** run on `from_slot` and:
```
slot_meta.clear_unconfirmed_slot();
write_batch.put::<cf::SlotMeta>(slot, &slot_meta)?;
```
is **always** run on `to_slot`.

Reply:
but perhaps this is out of scope for this function and we should only `run_purge` for rooted slots. in that case I agree with carl's idea of separating this out to a different function that is intended for forks currently being replayed.